ipld-importer/add.go

```diff
@@ -0,0 +1,439 @@
+package importer
+
+import (
+	"context"
+	"fmt"
+	"io"
+	gopath "path"
+	"strconv"
+
+	"github.com/ipfs/ipfs-cluster/api"
+
+	cid "github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	syncds "github.com/ipfs/go-datastore/sync"
+	files "github.com/ipfs/go-ipfs-cmdkit/files"
+	bstore "github.com/ipfs/go-ipfs/blocks/blockstore"
+	bserv "github.com/ipfs/go-ipfs/blockservice"
+	"github.com/ipfs/go-ipfs/exchange/offline"
+	balanced "github.com/ipfs/go-ipfs/importer/balanced"
+	"github.com/ipfs/go-ipfs/importer/chunk"
+	ihelper "github.com/ipfs/go-ipfs/importer/helpers"
+	trickle "github.com/ipfs/go-ipfs/importer/trickle"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	mfs "github.com/ipfs/go-ipfs/mfs"
+	"github.com/ipfs/go-ipfs/pin"
+	posinfo "github.com/ipfs/go-ipfs/thirdparty/posinfo"
+	unixfs "github.com/ipfs/go-ipfs/unixfs"
+	ipld "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log"
+)
+
+var log = logging.Logger("coreunix")
+
+// how many bytes of progress to wait before sending a progress update message
+const progressReaderIncrement = 1024 * 256
+
+var liveCacheSize = uint64(256 << 10)
+
+type hiddenFileError struct {
+	fileName string
+}
+
+func (e *hiddenFileError) Error() string {
+	return fmt.Sprintf("%s is a hidden file", e.fileName)
+}
+
+type ignoreFileError struct {
+	fileName string
+}
+
+func (e *ignoreFileError) Error() string {
+	return fmt.Sprintf("%s is an ignored file", e.fileName)
+}
+
+// NewAdder Returns a new Adder used for a file add operation.
+func NewAdder(ctx context.Context, p pin.Pinner, bs bstore.GCBlockstore, ds ipld.DAGService) (*Adder, error) {
```

Comment:
Is this a simplified copy from the ipfs adder?

Reply:
Yes!  I cut away all of the GC and pinning stuff.  It also has some slight tweaks in the code that outputs printable objects for reporting progress.  I'll make a document level comment describing the source and the tweaks from the original.